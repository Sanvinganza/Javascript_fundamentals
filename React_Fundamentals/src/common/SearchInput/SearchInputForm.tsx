import { useState, useCallback, useMemo } from 'react';
import { SearchInput } from './SearchInput';
import { ESearchInputModes as SearchInputModes} from './SearchInput';

export interface SearchInputFormProps {
  array: Array<string>
}

export function SearchInputForm({array}:SearchInputFormProps) {
  const [searchResult, setSearchResult] = useState(array);
  const [mode, setMode] = useState(SearchInputModes.withDelay);

  const onSearchHandler = useCallback((value) => {
    setSearchResult(() => array.filter(el => el.includes(value)));
  }, [array]);

  const memoArrayResult = useMemo(() => searchResult.map(result => {
    return (<p key={result}>{result}</p>);
  }), [searchResult]);

  return (
    <>
      <SearchInput
        placeholder="Search"
        mode={mode}
        onSearch={onSearchHandler}
      />
      <form>
        <label htmlFor="immediate">immediate</label>
        <input data-testid="immediate" id="immediate" type='radio' name="mode" onClick={() =>
          setMode(SearchInputModes.immediate)
        } />

        <label htmlFor="withDelay">withDelay</label>
        <input data-testid="withDelay" id="withDelay" type='radio' name="mode" onClick={() =>
          setMode(SearchInputModes.withDelay)
        } />

        <label htmlFor="onPress">onPress</label>
        <input data-testid="onPress" id="onPress" type='radio' name="mode" onClick={() =>
          setMode(SearchInputModes.onPress)
        } />
      </form>
      {memoArrayResult}
    </>
  );
}
