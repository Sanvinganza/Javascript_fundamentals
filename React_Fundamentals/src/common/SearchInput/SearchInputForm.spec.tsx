import { screen, render, waitFor } from "@testing-library/react"
import { SearchInputForm } from "./SearchInputForm";
import userEvent from "@testing-library/user-event";
import { keyboard } from "@testing-library/user-event/dist/keyboard";
import { fireEvent } from "@testing-library/react";

describe('SearchInputForm component', () => {
  const array = ['word', 'noun', 'hello', 'world'];

  const setup = () => {
    const utils = render(<SearchInputForm array={array}/>);
    const input = utils.getByPlaceholderText('Search');

    return {
      input,
      ...utils
    }
  }

  test('renders corrently', () => {
    expect(render(<SearchInputForm array={array} />));
  })
  test('input should be work with Immediate options currently', () => {
    const { input, asFragment, getByDisplayValue } = setup();

    expect(input).toBeInTheDocument();

    const inputImmediate = screen.getByTestId ('immediate');
    expect(inputImmediate).toBeInTheDocument();

    userEvent.click(inputImmediate);
    expect(inputImmediate).toBeChecked();

    userEvent.click(input);
    keyboard('w');
    expect(getByDisplayValue('w')).toEqual(input);
    
    expect(screen.getAllByRole('listitem').length).toEqual(2);
    expect(asFragment()).toMatchSnapshot();
  })
  test('input should be work with withDelay', async () => {
    const { input, asFragment, getByDisplayValue } = setup();

    expect(input).toBeInTheDocument();

    const inputWithDelay = screen.getByTestId ('withDelay');
    expect(inputWithDelay).toBeInTheDocument();

    userEvent.click(inputWithDelay);
    expect(inputWithDelay).toBeChecked();

    userEvent.click(input);
    keyboard('or');
    await waitFor(() => expect(getByDisplayValue('or')).toEqual(input));
    
    await waitFor(() => expect(screen.getAllByRole('listitem').length).toEqual(2));
    expect(asFragment()).toMatchSnapshot();
  })
  test('input should be work with onPress', () => {
    const { input, asFragment } = setup();

    expect(input).toBeInTheDocument();

    const inputOnPress = screen.getByTestId ('onPress');
    expect(inputOnPress).toBeInTheDocument();

    userEvent.click(inputOnPress);
    expect(inputOnPress).toBeChecked();

    userEvent.click(input);
    userEvent.type(input, 'w');
    fireEvent.keyDown(input, {key: 'Enter', code: 'Enter', charCode: 13})
    
    expect(screen.getAllByRole('listitem').length).toEqual(2);
    expect(asFragment()).toMatchSnapshot();
  })
})