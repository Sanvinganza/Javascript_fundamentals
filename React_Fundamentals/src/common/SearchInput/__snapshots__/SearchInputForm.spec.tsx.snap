// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInputForm component input should be work with Immediate options currently 1`] = `
<DocumentFragment>
  <form>
    <fieldset>
      <legend>
        Search Input
      </legend>
      <input
        aria-label="search-input"
        id="search-input"
        placeholder="Search"
        type="search"
      />
    </fieldset>
  </form>
  <form>
    <label
      for="immediate"
    >
      immediate
    </label>
    <input
      data-testid="immediate"
      id="immediate"
      name="mode"
      type="radio"
    />
    <label
      for="withDelay"
    >
      withDelay
    </label>
    <input
      data-testid="withDelay"
      id="withDelay"
      name="mode"
      type="radio"
    />
    <label
      for="onPress"
    >
      onPress
    </label>
    <input
      data-testid="onPress"
      id="onPress"
      name="mode"
      type="radio"
    />
  </form>
  <ul>
    <li>
      word
    </li>
    <li>
      world
    </li>
  </ul>
</DocumentFragment>
`;

exports[`SearchInputForm component input should be work with onPress 1`] = `
<DocumentFragment>
  <form>
    <fieldset>
      <legend>
        Search Input
      </legend>
      <input
        aria-label="search-input"
        id="search-input"
        placeholder="Search"
        type="search"
      />
    </fieldset>
  </form>
  <form>
    <label
      for="immediate"
    >
      immediate
    </label>
    <input
      data-testid="immediate"
      id="immediate"
      name="mode"
      type="radio"
    />
    <label
      for="withDelay"
    >
      withDelay
    </label>
    <input
      data-testid="withDelay"
      id="withDelay"
      name="mode"
      type="radio"
    />
    <label
      for="onPress"
    >
      onPress
    </label>
    <input
      data-testid="onPress"
      id="onPress"
      name="mode"
      type="radio"
    />
  </form>
  <ul>
    <li>
      word
    </li>
    <li>
      world
    </li>
  </ul>
</DocumentFragment>
`;

exports[`SearchInputForm component input should be work with withDelay 1`] = `
<DocumentFragment>
  <form>
    <fieldset>
      <legend>
        Search Input
      </legend>
      <input
        aria-label="search-input"
        id="search-input"
        placeholder="Search"
        type="search"
      />
    </fieldset>
  </form>
  <form>
    <label
      for="immediate"
    >
      immediate
    </label>
    <input
      data-testid="immediate"
      id="immediate"
      name="mode"
      type="radio"
    />
    <label
      for="withDelay"
    >
      withDelay
    </label>
    <input
      data-testid="withDelay"
      id="withDelay"
      name="mode"
      type="radio"
    />
    <label
      for="onPress"
    >
      onPress
    </label>
    <input
      data-testid="onPress"
      id="onPress"
      name="mode"
      type="radio"
    />
  </form>
  <ul>
    <li>
      word
    </li>
    <li>
      world
    </li>
  </ul>
</DocumentFragment>
`;
